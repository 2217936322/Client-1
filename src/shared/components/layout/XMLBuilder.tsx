import React, { useMemo, useState, useEffect } from 'react';
import parser, { iterativeConfigKey } from './parser/xml-parser';
import * as processor from './processor';
import _isEmpty from 'lodash/isEmpty';
import _isUndefined from 'lodash/isUndefined';
import _isNil from 'lodash/isNil';
import './tags/__all__';
import styled from 'styled-components';
import { StateDataType } from './types';
import { useSize } from 'react-use';
import { LayoutWidthContextProvider } from './context/LayoutWidthContext';
import { LayoutStateContextProvider } from './context/LayoutStateContext';
import { useBuildLayoutStateContext } from './hooks/useBuildLayoutStateContext';
import { TagComponent } from './tags/type';

/**
 * XML布局需求:
 * name必须全局唯一
 * 需要版本号
 * 以Template组件作为统一包装(不强求)
 */

export type DefinePropsType = {
  [name: string]: any;
};
interface DefineMap {
  [name: string]: TagComponent;
}

export interface DataMap {
  _name?: string;
  _avatar?: string;
  _desc?: string;
  [name: string]: StateDataType | DataMap; // 可以嵌套
}

export type LayoutType = 'edit' | 'detail';

export interface XMLBuilderState {
  defines: DefineMap;
  data: DataMap;
}

export interface XMLBuilderAction {
  type: string;
  payload: { [name: string]: any };
  [others: string]: any;
}

export interface XMLBuilderContext {
  state: XMLBuilderState;
  dispatch: React.Dispatch<XMLBuilderAction>;
  layoutType: LayoutType;
  parent?: XMLBuilderContext;
}

export type StateChangeHandler = (newState: XMLBuilderState) => void;

const XMLBuilderContainer = styled.div`
  text-align: left;

  pre {
    margin: 0;
  }
`;

class XMLErrorBoundary extends React.Component {
  state = { hasError: false, error: null };

  static getDerivedStateFromError(error) {
    // 更新 state 使下一次渲染能够显示降级后的 UI
    return { hasError: true, error };
  }

  componentDidCatch(error, errorInfo) {
    // 你同样可以将错误日志上报给服务器
    console.error(error, errorInfo);
  }

  render() {
    if (this.state.hasError) {
      // 你可以自定义降级后的 UI 并渲染
      return (
        <div>
          <h2>人物卡布局解析错误</h2>
          <p>{String(this.state.error)}</p>
        </div>
      );
    }

    return this.props.children;
  }
}

interface Props {
  xml: string;
  layoutType?: LayoutType;
  initialData?: DataMap;
  onChange?: StateChangeHandler;
}
const XMLBuilder: React.FC<Props> = React.memo((props) => {
  const { xml = '', onChange, layoutType = 'edit' } = props;
  const [error, setError] = useState<Error>(null);
  const [layout, setLayout] = useState();

  const { state, dispatch } = useBuildLayoutStateContext({
    initialData: props.initialData,
    onChange,
  });

  useEffect(() => {
    try {
      const layout = parser(xml);
      layout.type = 'root';
      iterativeConfigKey(layout);
      console.log('layout', layout);
      setLayout(layout);
      setError(null);
    } catch (err) {
      console.error(err);
      setError(err);
    }
  }, [xml]);

  const LayoutDOM = useMemo(() => {
    if (!_isNil(error)) {
      return (
        <div>
          <p>XML解析出现错误:</p>
          <p>{String(error)}</p>
        </div>
      );
    }

    if (_isEmpty(layout)) {
      return null;
    }

    try {
      return processor.render(layout, { state, dispatch, layoutType });
    } catch (err) {
      console.error(err);
      setError(err);
      return null;
    }
  }, [layout, state, dispatch, layoutType, error]);

  const [XMLRender, { width }] = useSize(
    <XMLBuilderContainer>{LayoutDOM}</XMLBuilderContainer>
  );

  return (
    <XMLErrorBoundary>
      <LayoutWidthContextProvider width={width}>
        <LayoutStateContextProvider state={{ state, dispatch, layoutType }}>
          {XMLRender}
        </LayoutStateContextProvider>
      </LayoutWidthContextProvider>
    </XMLErrorBoundary>
  );
});
XMLBuilder.displayName = 'XMLBuilder';

export default XMLBuilder;
