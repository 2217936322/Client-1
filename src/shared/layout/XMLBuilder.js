import React, { useEffect, useState } from 'react';
import parser from './parser';
import processor from './processor';
import _isEmpty from 'lodash/isEmpty';

const XMLBuilder = (props) => {
  const { xml = '' } = props;
  const [layout, setLayout] = useState({});

  useEffect(
    () => {
      const data = parser(xml);
      data.type = 'root';
      console.log('data', data);
      setLayout(data);
    },
    [xml]
  );

  if (_isEmpty(layout)) {
    return null;
  }

  return <div>xmlbuilder:{processor.render(layout)}</div>;
};

export default XMLBuilder;
