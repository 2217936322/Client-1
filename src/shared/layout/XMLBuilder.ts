import React, { useEffect, useState, useContext } from 'react';
import parser from './parser';
import processor from './processor';
import _isEmpty from 'lodash/isEmpty';
import _isUndefined from 'lodash/isUndefined';
import './types/__all__';

const XMLBuilder = (props) => {
  const { xml = '' } = props;
  const [layout, setLayout] = useState({});
  const context = React.createContext(null);

  useEffect(
    () => {
      const data = parser(xml);
      data.type = 'root';
      console.log('data', data);
      setLayout(data);
    },
    [xml]
  );

  if (_isEmpty(layout) || _isUndefined(context)) {
    return null;
  }

  const Provider = context.Provider;
  return (
    <div>
      xmlbuilder:
      <Provider>{processor.render(layout, context)}</Provider>
    </div>
  );
};

export default XMLBuilder;
