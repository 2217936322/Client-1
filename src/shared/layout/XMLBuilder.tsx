import React, { useEffect, useState, useReducer } from 'react';
import parser, { XMLElement } from './parser/xml-parser';
import * as processor from './processor';
import _clone from 'lodash/clone';
import _isEmpty from 'lodash/isEmpty';
import _isUndefined from 'lodash/isUndefined';
import './types/__all__';
import Debug from 'debug';
import styled from 'styled-components';
const debug = Debug('trpg:XMLBuilder');

export type DefinePropsType = {
  [name: string]: any;
};
interface DefineType {
  [name: string]: (
    context: XMLBuilderContext,
    otherProps: DefinePropsType
  ) => React.FunctionComponentElement<{}>;
}

interface GlobalType {
  [name: string]: number | string | null;
}

export interface DataType {
  [name: string]: number | string | null;
}

export type LayoutType = 'edit' | 'detail';

export interface XMLBuilderState {
  defines: DefineType;
  global: GlobalType;
  data: DataType;
}

export interface XMLBuilderAction {
  type: string;
  payload: { [name: string]: any };
  [others: string]: any;
}

export interface XMLBuilderContext {
  state: XMLBuilderState;
  dispatch: React.Dispatch<XMLBuilderAction>;
  layoutType: LayoutType;
}

type stateChangeHandler = (newState: XMLBuilderState) => void;

export enum ActionType {
  UpdateData = 'update_data',
  AddDefine = 'add_define',
}

const XMLBuilderContainer = styled.div`
  text-align: left;
`;

const buildReducer = (onChange?: stateChangeHandler) => {
  const XMLBuilderReducer = (
    prevState: XMLBuilderState,
    action: XMLBuilderAction
  ): XMLBuilderState => {
    const type = action.type;
    const payload = action.payload;
    const newState = _clone(prevState);
    debug(`[Action] ${type}: %o`, payload);

    switch (type) {
      case ActionType.UpdateData: {
        const scope = action.scope || 'data';
        newState[scope] = payload;
        break;
      }
      case ActionType.AddDefine:
        newState.defines[payload.name] = payload.componentFn;
        break;
    }

    onChange && onChange(newState);

    return newState;
  };

  return XMLBuilderReducer;
};

interface Props {
  xml: string;
  layoutType?: LayoutType;
  initialData?: DataType;
  onChange?: stateChangeHandler;
}
const XMLBuilder = (props: Props) => {
  const { xml = '', onChange, layoutType = 'edit' } = props;
  const [layout, setLayout] = useState({});

  const initialState: XMLBuilderState = {
    defines: {},
    global: {},
    data: props.initialData || {},
  };
  const [state, dispatch] = useReducer(buildReducer(onChange), initialState);

  useEffect(() => {
    const layout = parser(xml);
    layout.type = 'root';
    console.log('layout', layout);
    setLayout(layout);
  }, [xml]);

  if (_isEmpty(layout)) {
    return null;
  }

  return (
    <XMLBuilderContainer>
      {processor.render(layout, { state, dispatch, layoutType })}
    </XMLBuilderContainer>
  );
};

export default XMLBuilder;
