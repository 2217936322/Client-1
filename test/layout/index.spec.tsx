import React from 'react';
import XMLBuilder from '@shared/components/layout/XMLBuilder';
import { shallow, mount } from 'enzyme';

const wrapXML = (xml: string) => `<Template>${xml}</Template>`;
const renderXML = (xml: string) => shallow(<XMLBuilder xml={wrapXML(xml)} />);
// const mountXML = (xml: string) => mount(<XMLBuilder xml={wrapXML(xml)} />); // 使用会抛个bug，暂时不要用

describe('layout', () => {
  test('basic', () => {
    const wrapper = renderXML('');
    expect(wrapper.children()).toHaveLength(1);
    expect(wrapper.toJson()).toMatchSnapshot();
  });

  test('some text', () => {
    const text = '<span>some text</span>';
    const wrapper = renderXML(text);

    expect(
      wrapper
        .children()
        .children()
        .html()
    ).toBe(text);
    expect(wrapper.toJson()).toMatchSnapshot();
  });

  test.only('layout should update when xml change', () => {
    const text = '<span>some text</span>';
    const text2 = '<span>some text2</span>';
    const wrapper = renderXML(text);

    expect(
      wrapper
        .children()
        .children()
        .html()
    ).toBe(text);
    expect(wrapper.toJson()).toMatchSnapshot();

    wrapper.setProps({ xml: wrapXML(text2) });

    expect(
      wrapper
        .children()
        .children()
        .html()
    ).toBe(text2);
    expect(wrapper.toJson()).toMatchSnapshot();
  });
});
