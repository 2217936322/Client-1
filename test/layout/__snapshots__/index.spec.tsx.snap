// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`layout layout should update when xml change 1`] = `
<styled.div>
  <styled.div>
    <span
      key="Template-span-0"
    >
      some text
    </span>
  </styled.div>
</styled.div>
`;

exports[`layout layout should update when xml change 2`] = `
<styled.div>
  <styled.div>
    <span
      key="Template-span-0"
    >
      some text2
    </span>
  </styled.div>
</styled.div>
`;

exports[`layout some text 1`] = `
<styled.div>
  <styled.div>
    <span
      key="Template-span-0"
    >
      some text
    </span>
  </styled.div>
</styled.div>
`;
